tools/viewer.py

```diff
@@ -0,0 +1,375 @@
+    shouldRefreshNow = False
+    event = pygame.event.wait()
+    if event.type == QUIT: return
+    if event.type == KEYDOWN:
+      if event.key in (K_ESCAPE,K_q): return 
+      if event.key == K_p:
+        if index + 1 < len( posesScanSet ):
+          index += 1
+      if event.key == K_o:
+        if index + 10 < len( posesScanSet ):
+          index += 10
+      if event.key == K_i:
+        if index > 0:
+          index -= 1
+      if event.key == K_0:
+        index = 0
+      if event.key == K_9:
+        index = len( posesScanSet ) - 1
+      if event.key == K_m:
+        shouldDrawMap = not shouldDrawMap;        
+      if event.key == K_s:
+        shouldDrawSensors = not shouldDrawSensors;        
+      if event.key == K_b:
+        shouldDrawBeacons = not shouldDrawBeacons;        
+      if event.key == K_RIGHT:
+        globals()['offset'] = (offset[0]+150, offset[1])
+        shouldRefreshNow = True
+      if event.key == K_LEFT:
+        globals()['offset'] = (offset[0]-150, offset[1])
+        shouldRefreshNow = True
+      if event.key == K_UP:
+        globals()['offset'] = (offset[0], offset[1]+150)
+        shouldRefreshNow = True
+      if event.key == K_DOWN:
+        globals()['offset'] = (offset[0], offset[1]-150)
+        shouldRefreshNow = True
+      if event.key == K_PLUS or event.unicode == '+':
+        globals()['scale'] *= 2.0
+        globals()['tile_size'] /= 2.0
+        shouldDrawTiles = shouldDrawMap = shouldDrawSensors = shouldDrawBeacons = True
+        shouldRefreshNow = True
+      if event.key == K_MINUS or event.unicode == '-':
+        globals()['scale'] /= 2.0
+        globals()['tile_size'] *= 2.0
+        shouldDrawTiles = shouldDrawMap = shouldDrawSensors = shouldDrawBeacons = True
+        shouldRefreshNow = True
+
+      def has_image(k):
+        return bool(posesScanSet[k][2])
+
+#      if True: #hack has_image(index) or shouldRefreshNow:
+      if has_image(index) or shouldRefreshNow:
+        foreground.fill( (0,0,0) )
+
+      refreshed_since = index
+
+      if shouldRefreshNow:
+        while refreshed_since > 0:
+          if has_image(refreshed_since):
+            break
+          refreshed_since -= 1
+
+      for i in range(refreshed_since, index + 1):
+        drawPoses( foreground, posesScanSet[i][0] )
+        drawCompass( foreground, posesScanSet[i] )
+        drawScans( foreground, posesScanSet[i][1], shouldDrawSensors, shouldDrawBeacons )
+        drawImage( foreground, posesScanSet[i][2], posesScanSet[i][3] )
+        imgFileName = posesScanSet[i][2]
+
+      if event.key == K_a:
+        foreground.fill( (0,0,0) )
+        for (p,s,i,camdir,comp) in posesScanSet:
+          drawScans( foreground, s, shouldDrawSensors, shouldDrawBeacons )
+
+    if shouldDrawTiles:
+      drawTiles(background)
+
+    if shouldDrawMap:
+      drawMap( foreground, map )
+    screen.blit(background, (0, 0)) 
+    screen.blit(foreground, (0, 0))
+    pygame.display.flip() 
```

Comment:
I would postpone changes to `viewer.py` to later stage (I expect large refactoring).